### PR TITLE
ci: reorganize github CI tests

### PR DESCRIPTION
**Pull request by mr-cal**
- [X] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?

-----

## Overview

Reorganize github CI tests and workaround failing MacOS integration tests.

## Details

1. Separate linters, unit tests, and integration tests
2. Align with starbase's [workflow](https://github.com/canonical/starbase/blob/main/.github/workflows/tests.yaml) (this PR is a first step, not the end result)
3. Add more unit testing across OS's and Python versions 

(CRAFT-1726)

### CODE REVIEW

File: .github/workflows/tests.yaml [hunk 1]
```diff
@@ -46,70 +59,93 @@ jobs:
...
-          sudo snap install --classic node
-          sudo snap install --classic pyright
           make test-pyright
-      - name: Install LXD
+
+  unit-tests:
+    strategy:
+      matrix:
+        os: [ubuntu-latest, macos-latest, windows-latest]
+        python-version: ["3.8", "3.9", "3.10"]
```
Comment (lengau) [inline .github/workflows/tests.yaml:68]: Do we need to test on 3.9? I'd kinda prefer to do 3.8, 3.10, 3.11 if possible, getting us a [focal, jammy, latest] support matrix.

File: .github/workflows/tests.yaml [hunk 2]
```diff
@@ -46,70 +59,93 @@ jobs:
...
+          fetch-depth: 0
+      - name: Set up Python versions on ${{ matrix.platform }}
+        uses: actions/setup-python@v4
+        with:
+          python-version: ${{ matrix.python-version }}
+      - name: Configure environment
+        run: |
+          echo "::group::pip install"
+          python -m pip install -U .[dev]
+          python -m pip install -U -e .
```
Comment (lengau) [inline .github/workflows/tests.yaml:83]:
```suggestion
          python -m pip install -U -e .[dev]
```

will do the same thing, right?

File: .github/workflows/tests.yaml [hunk 3]
```diff
@@ -46,70 +59,93 @@ jobs:
...
+          echo "::endgroup::"
+      - name: Run unit tests
+        run: make test-units
+
+  integration-tests-linux:
+    # TODO: does not work on ubuntu-latest or with canonical/setup-lxd
```
Comment (lengau) [inline .github/workflows/tests.yaml:89]: Can you create (and reference in the comment) an issue about this?

File: .github/workflows/tests.yaml [hunk 4]
```diff
@@ -46,70 +59,93 @@ jobs:
...
+    runs-on: ubuntu-20.04
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v3
+        with:
+          fetch-depth: 0
+      - name: Setup Python
+        uses: actions/setup-python@v4
+        with:
+          python-version: '3.10'
```
Comment (lengau) [inline .github/workflows/tests.yaml:99]: Can we do a matrix of Python `["3.8", "3.10", "3.11"]` on this too?

**Review by lengau — APPROVED** [review]
Looks good! Picked some nits

File: .github/workflows/tests.yaml (hunk 1 above)
Comment (mr-cal) [reply .github/workflows/tests.yaml]: I removed 3.9 and created an issue for 3.11 support.

File: .github/workflows/tests.yaml (hunk 2 above)
Comment (mr-cal) [reply .github/workflows/tests.yaml:84]: Yes, fixed!

File: .github/workflows/tests.yaml (hunk 3 above)
Comment (mr-cal) [reply .github/workflows/tests.yaml]: Done!

File: .github/workflows/tests.yaml (hunk 4 above)
Comment (mr-cal) [reply .github/workflows/tests.yaml]: Done!

**Pull request merged**